### PR TITLE
FIP-0064: Add Parameters to Base Fee Mechanism.md

### PR DESCRIPTION
**Pull request by guy-goren**
Proposing only the first part from the discussion agenda (with the same name). That is, the weighted averaging.

### CODE REVIEW

File: FIPS/fip-xxx.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
+
```
Comment (anorth) [inline FIPS/fip-xxx.md:1]: Please include the standard headers table from the template.

File: FIPS/fip-xxx.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+The current formula for setting the base fee follows the original EIP1559:
+$$b[i+1]\triangleq  b[i] \cdot \left( 1+\frac{1}{8} \cdot \frac{s[i]-s^* }{s^* }\right)$$ where $i$ enumerates the epoch, $s^*$ is a prederemined block size (the "desired" size), $b[i]$ is the base fee at epoch $i$, and $s[i]$ is the block size at epoch $i$. This formula considers only the block size $s[i]$ from the last epoch. This mechanism might lead to incentives for users to bribe miners in order to reduce the base fee, or (in analogous manner) for miners to intiate a collusion with sophisticated users for the benefit of both. (See [next section](https://hackmd.io/pAgUi8mxRy-eQ7LNIIFnkQ#Background-and-Motivation).)
+
+We propose to consider the history of block sizes via a wheighted average with a geometric sequence as the weights. In particular, we suggest the following update rule:
+$$b[i+1]\triangleq  b[i] \cdot \left( 1+\frac{1}{8} \cdot \frac{s_{\textit{avg}}[i]-s^* }{s^* }\right)$$ where $s_{\textit{avg}}[i]$ is defined by
+$$s_{\textit{avg}}[i] \triangleq \frac{1-q}{q}\sum_{k=1}^{\infty} q^k\cdot s[i-k+1], \hspace{1cm}   q\in (0,1).$$ 
+
+## Background and Motivation
```
Comment (anorth) [inline FIPS/fip-xxx.md:14]: Please move the body down to the Change Motivation heading and delete this one.

File: FIPS/fip-xxx.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
+
+# FIP: Add Parameters to Base Fee Mechanism
+The first part of [this discussion](https://github.com/filecoin-project/FIPs/discussions/686)
+## Simple Summary
```
Comment (anorth) [inline FIPS/fip-xxx.md:4]: Please follow the template to include
- Simple summary (1-2 sentences)
- Abstract (~200 words)

File: FIPS/fip-xxx.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+For simplicity purposes we will describe the problem (and proposed improvements) in the simplified Ethereum model of a single chain of blocks. While Filecoin has multiple blocks in a single epoch, the same principles apply.
```
Comment (anorth) [inline FIPS/fip-xxx.md:5]: This was fine for discussion, but for the formal FIP please describe in Filecoin terms, including tipset details where they are relevant.

File: FIPS/fip-xxx.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+The current formula for setting the base fee follows the original EIP1559:
+$$b[i+1]\triangleq  b[i] \cdot \left( 1+\frac{1}{8} \cdot \frac{s[i]-s^* }{s^* }\right)$$ where $i$ enumerates the epoch, $s^*$ is a prederemined block size (the "desired" size), $b[i]$ is the base fee at epoch $i$, and $s[i]$ is the block size at epoch $i$. This formula considers only the block size $s[i]$ from the last epoch. This mechanism might lead to incentives for users to bribe miners in order to reduce the base fee, or (in analogous manner) for miners to intiate a collusion with sophisticated users for the benefit of both. (See [next section](https://hackmd.io/pAgUi8mxRy-eQ7LNIIFnkQ#Background-and-Motivation).)
```
Comment (anorth) [inline FIPS/fip-xxx.md:8]: Please fix link to "next section" to be relative rather than to hackmd, or just omit it.

File: FIPS/fip-xxx.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+Research on the subject (focused on ethereum) have revealed issues with this transaction fee mechanism. It has been shown to be [unstable in cases](https://arxiv.org/pdf/2102.10567.pdf). Moreover, it has been shown that the dynamic nature of the base fee, which is influenced by the fill rate of blocks, opens the door for [manipulation by miners and users](https://drive.google.com/file/d/1o_47aiN1BWegOmG8utYh4a47TM0OOa12/view?usp=share_link). The desired behavior of the system under a stable high demand, is for it to reach an equalibrium where the base fee -- $b$ -- is the significant part of the gas fee, and the tip is relatively small -- denoted $\varepsilon$ (for reference, Ethereum often has $\frac{b}{\varepsilon}\approx 20$). According to [Roughgarden](https://timroughgarden.org/papers/eip1559.pdf) this is a rational equalibrium under the assumption that miners do not think ahead. However, we expect a miner to optimize its behavior by also considering its future payoffs. In essence, since neither the miner nor the user are getting the burnt fee, by colluding they can both tap into the burnt fee for a win-win situation for them both.
```
Comment (anorth) [inline FIPS/fip-xxx.md:19]: Is it possible to get a more permanent link/citation than a PDF on Google Drive for the second one?

File: FIPS/fip-xxx.md [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+A [theorethical work](https://drive.google.com/file/d/1o_47aiN1BWegOmG8utYh4a47TM0OOa12/view?usp=share_link) describes how both miners and users can initiate an attack. For example, we can imagine that users who wish to pay lower costs will coordinate the attack. Roughly, a user (or group of users) that has transactions with a total $g$ amount of gas bribes the miner of the current block (no matter the miner's power) to propose an empty block instead. The cost of such a bribe is only $\varepsilon \times {s^* }$ -- the tip times the target block size. Consequently, the base fee reduces in the next block. If we accept that EIP1559 reaches its goals, e.g., users would typicaly use a simple and honest bidding strategy of reporting their maximal willingness to pay plus adding a small tip ($\varepsilon$), then in the honest users' steady state, gas proposals leave the miners with an $\varepsilon$ tip. Given that other users are na\"ive (or slow to react), our bribing user will include its transactions with any tip larger than $\varepsilon$ -- making the attack profitable whenever $g \frac{b^* }{8} >s^* \varepsilon$.
```
Comment (anorth) [inline FIPS/fip-xxx.md:21]: Ditto, can we link to a more permanent resource than someone's Drive?

Edit: I see this is the same link, we probably only need it once.

File: FIPS/fip-xxx.md [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+The [full suggestion](https://github.com/filecoin-project/FIPs/discussions/686) is actually composed of two parts: (i) using an average over epochs of block sizes rather than only the latest, and (ii) using an adaptive learning rate. The first part is the basis of the proposal. The second is only an additional improvement (that depends on the first part being implemented). This FIP, therefore, discusses only the first part and leaves the second part for a later time.
+
+### $s_\textit{avg}[i]$ instead of $s[i]$
+The change is based on [this work](https://drive.google.com/file/d/1o_47aiN1BWegOmG8utYh4a47TM0OOa12/view?usp=share_link) that described a rational strategy in which bribes are profitabe. Choosing to average based on a geometric series weights results in two desired properties: (i) the computation and space complexity are both in O(1), and (ii) the average gradually phases out the impact of a single outlier block without causing significant fluctuations in the base fee.
```
Comment (anorth) [inline FIPS/fip-xxx.md:58]: Same GDrive link again here.

Is the rational bribe strategy concise enough to just include inline here?

File: FIPS/fip-xxx.md [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+The proposal is designed to improve the incentive compatability of the TFM. A [game theoretic analysys](https://drive.google.com/file/d/1o_47aiN1BWegOmG8utYh4a47TM0OOa12/view?usp=share_link) shows that a TFM based on EIP-1559 encourages bribes. Roughly, because the base fee in the next block depends on the size of the the current block, a miner that creates an empty block reduces the base fee of the next block by a factor of $\frac{1}{8}$. The opportunity cost for mining an empty block instead of a normal block is only the tips it contains. Thus, the cost of bribing a miner is only compansating it for the lost tips. In case the base fee is significantly larger than the tips, the bribing user gains a siginificant reduction in the base fees of the next block, making the bribe profitable.
```
Comment (anorth) [inline FIPS/fip-xxx.md:73]: The last sentences here are a great short description of the problem. Consider promoting it to the change motivation section.

File: FIPS/fip-xxx.md [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+One of the main goals of EIP-1559 was to simplify the bidding for users. It was articulated [theoreticaly by Roughgarden](https://timroughgarden.org/papers/eip1559.pdf) as users bidding their honest valuations being an optimal strategy. In contrast, when using first price auctions for the TFM (as done by Bitcoin and previously in Ethereum), it is typically sub-optimal for a user to bid its honest valuation. In other words, a TFM that encourages users to not fully reveal their preferences is considerd less good. However, our opinion is that a TFM that encourages bribes is worse than a TFM that encourages not revealing one's full preferences.
+
+We belive that a first price auction is the best way to go regarding TFMs, however, Filecoin chooses to use EIP-1559 and burn transaction fees (perhaps for reasons other than game-theoretic ones). We therefore suggest to mitigate the current incentives for bribes using the above proposal.
```
Comment (anorth) [inline FIPS/fip-xxx.md:77]: Nit: try to remove the language like "we", "opinion", and just state the proposal and the relevant considerations and tradeoffs objectively. Your view about first-price auctions isn't improving the spec here.

**Comment by anorth** [comment]
Please use FIP number 0063.

File: FIPS/fip-xxx.md (hunk 1 above)
Comment (guy-goren) [reply FIPS/fip-xxx.md]: done

File: FIPS/fip-xxx.md (hunk 2 above)
Comment (guy-goren) [reply FIPS/fip-xxx.md]: done

File: FIPS/fip-xxx.md (hunk 3 above)
Comment (guy-goren) [reply FIPS/fip-xxx.md:13]: done

File: FIPS/fip-xxx.md (hunk 5 above)
Comment (guy-goren) [reply FIPS/fip-xxx.md]: done

File: FIPS/fip-xxx.md (hunk 6 above)
Comment (guy-goren) [reply FIPS/fip-xxx.md]: done

File: FIPS/fip-xxx.md (hunk 9 above)
Comment (guy-goren) [reply FIPS/fip-xxx.md]: I added a "pointer" to them.

File: FIPS/fip-xxx.md (hunk 10 above)
Comment (guy-goren) [reply FIPS/fip-xxx.md]: done

File: FIPS/fip-0064.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
+---
+fip: "0064"
+title: Add Parameters to Base Fee Mechanism
+status: WIP
```
Comment (anorth) [inline FIPS/fip-0064.md:4]: This will be "Draft" as soon as we merge a PR, so just use that now.

**Review by anorth — APPROVED** [review]
Editorial approval as a draft.

File: FIPS/fip-0064.md [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+A [theorethical work](https://arxiv.org/abs/2304.11478) describes how both miners and users can initiate an attack. For example, we can imagine that users who wish to pay lower costs will coordinate the attack. Roughly, a user (or group of users) that has transactions with a total $g$ amount of gas bribes the miner of the current block (no matter the miner's power) to propose an empty block instead. The cost of such a bribe is only $\varepsilon \times {s^* }$ -- the tip times the target block size. Consequently, the base fee reduces in the next block. If we accept that EIP1559 reaches its goals, e.g., users would typicaly use a simple and honest bidding strategy of reporting their maximal willingness to pay plus adding a small tip ($\varepsilon$), then in the honest users' steady state, gas proposals leave the miners with an $\varepsilon$ tip. Given that other users are na\"ive (or slow to react), our bribing user will include its transactions with any tip larger than $\varepsilon$ -- making the attack profitable whenever $g \frac{b^* }{8} >s^* \varepsilon$.
```
[the remark this answers is not in the document]
Comment (kaitlin-beegle) [reply FIPS/fip-0064.md:34]: I know it's just a typo, but the word 'theorethical' seems WAY more interesting ;)

**Comment by kaitlin-beegle** [comment]
Just did a final editorial review and added a few grammar nits. Should be ok; going to merge.

File: FIPS/fip-0064.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+## Abstract
+The current formula for setting the base fee follows the original [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) (with minor adaptions to fit a tipset setting):
+$$b[i+1]\triangleq  b[i] \cdot \left( 1+\frac{1}{8} \cdot \frac{s[i]-s^* }{s^* }\right)$$ where $i$ enumerates the epoch, $s^*$ is a predetremined block size (the "desired" size), $b[i]$ is the base fee at epoch $i$, and $s[i]$ is the average size of a block in the tipset of epoch $i$. This formula considers only the average block size $s[i]$ from the last epoch. This mechanism might lead to incentives for users to bribe miners in order to reduce the base fee or, in an analogous manner, for miners to collude with sophisticated users for the benefit of both. (See Motivation section.)
+
+We propose to consider the history of block sizes via a time weighted average with a geometric sequence as the weights. In particular, we suggest the following update rule:
+$$b[i+1]\triangleq  b[i] \cdot \left( 1+\frac{1}{8} \cdot \frac{s_{\textit{avg}}[i]-s^* }{s^* }\right)$$ where $s_{\textit{avg}}[i]$ is defined by
```
Comment (jennijuju) [inline FIPS/fip-0064.md:21]: can you please elaborate this series of #$%^&U* in English lol?

**Pull request merged**